spot_ros2_control/examples/src/noarm_squat.cpp

```diff
@@ -0,0 +1,133 @@
+        "joint_states", 10, std::bind(&NoarmSquat::joint_states_callback, this, std::placeholders::_1));
+    timer_ = create_wall_timer((std::chrono::milliseconds)(static_cast<int>(1000. / command_rate)),
+                               std::bind(&NoarmSquat::timer_callback, this));
+  }
+
+ private:
+  // For storing joint angles
+  std::vector<double> stand_joint_angles_;
+  std::vector<double> squat_joint_angles_;
+  std::vector<double> init_joint_angles_;
+  // For storing differences relevant for calculating the interpolation
+  std::vector<double> diff_squat_stand_;
+  std::vector<double> diff_squat_init_;
+  // Command we send to the robot
+  std_msgs::msg::Float64MultiArray command_;
+  // Parameters and info about state
+  SquatState squat_state_;
+  int points_per_motion_;
+  bool initialized_;
+  int count_;
+  size_t njoints_;
+  // RCLCPP publishers, subscribers, timers
+  rclcpp::TimerBase::SharedPtr timer_;
+  rclcpp::Subscription<sensor_msgs::msg::JointState>::SharedPtr joint_states_sub_;
+  rclcpp::Publisher<std_msgs::msg::Float64MultiArray>::SharedPtr command_pub_;
+
+  void joint_states_callback(const sensor_msgs::msg::JointState& msg) {
+    // Used to save the starting joint angles
+    if (!initialized_) {
+      RCLCPP_INFO_STREAM(get_logger(), "Received starting joint states");
+      init_joint_angles_ = msg.position;
+      for (size_t i = 0; i < njoints_; i++) {
+        diff_squat_init_.push_back(squat_joint_angles_.at(i) - init_joint_angles_.at(i));
+      }
+      initialized_ = true;
+    }
+  }
+
+  void timer_callback() {
+    // Wait to send commands until we have initialized with the starting joint angles
+    if (!initialized_) {
+      return;
+    }
+    // Check if we need to switch state
+    if (count_ > points_per_motion_) {
+      switch (squat_state_) {
+        case INITIALIZING:
+          squat_state_ = SQUATTING;
+          break;
+        case SQUATTING:
+          squat_state_ = STANDING;
+          break;
+        case STANDING:
+          squat_state_ = SQUATTING;
+          break;
+      }
+      count_ = 0;
+    }
+    // Percentage we are through the desired motion
+    const double percentage = static_cast<float>(count_) / points_per_motion_;
+    // Fill in the command with the appropriate joint angles given the state
+    switch (squat_state_) {
+      case INITIALIZING:
+        for (size_t i = 0; i < njoints_; i++) {
+          command_.data.at(i) = percentage * diff_squat_init_.at(i) + init_joint_angles_.at(i);
+        }
+        break;
+      case SQUATTING:
+        for (size_t i = 0; i < njoints_; i++) {
+          command_.data.at(i) = percentage * -diff_squat_stand_.at(i) + squat_joint_angles_.at(i);
+        }
+        break;
+      case STANDING:
+        for (size_t i = 0; i < njoints_; i++) {
+          command_.data.at(i) = percentage * diff_squat_stand_.at(i) + stand_joint_angles_.at(i);
+        }
+        break;
+    }
```

Comment:
@khughes-bdai meta: same, much of this can be deduplicated with a helper function returning a reference to the corresponding baseline `*_joint_angles` based on state.

Reply:
Pushed a commit that modularizes this a lot with some helper functions and cleans up the actual timer callback